### PR TITLE
proxy args directly to ZSchema instance, allowing for choice between callback, promise, and sync apis

### PR DESCRIPTION
**Pull request by antialias**
This exposes zschema's promise and sync interface and preserves the existing cb interface.

The interesting change is in [`validator.js`](https://github.com/jsonresume/resume-schema/pull/356/files#diff-56729df6fb853f60216e1b2d64bccf91)

### CODE REVIEW

**Comment by antialias** [comment]
bump @evanplaice

**Comment by evanplaice** [comment]
2 things

`null` is used intentionally here to indicate that there is a value, it's just not a correctly typed valid value. `undefined` indicates the absence of a value. The tests can stay as-is.

The validator will soon be re-located to the `resume-validator` repo and linked into this project as an export to prevent backward-breaking changes.

We can break the behavior to ditch the callback pattern when 1.0 drops. It's a massive PITA but for a project that has this many consumers, it's important to avoid breaking current (even undesirable) behavior.

In short, ditch the changes to the tests. Submit changes to the validator under the [resume-validator](https://github.com/jsonresume/resume-validator) repo.

**Comment by antialias** [comment]
Sure, fine to close this PR since the validator and associated tests in this repo will be removed and published behind a major release that will guard existing code from the breaking change.

BTW, the offending `undefined`s come from `z-schema` so I put up a PR that changes `undefined` to `null` in their codebase where it matters, https://github.com/zaggino/z-schema/pull/254

> We can break the behavior to ditch the callback pattern when 1.0 drops.

The changes in this PR's branch don't ditch the callback pattern; both promise and callback are supported since `...args` is spread to the `validate` method.


>  It's a massive PITA but for a project that has this many consumers, it's important to avoid breaking current (even undesirable) behavior.

I count two dependencies that use the validator to check for errors, and they wouldn't be affected by the major release that this branch targets if we did "break" it:

* [`jsonresume-linkedin`](https://github.com/w0rd-driven/jsonresume-linkedin/blob/master/resume.js#L238-L251) does a truthiness check: `if (error || result.valid !== true) ...`
* [`theme-manager`](https://github.com/jsonresume/theme-manager/blob/master/lib/theme.js#L41-L46) passes the value to `stringify`: `JSON.stringify(err, undefined, 4)`

The rest of [`resume-schema`'s dependents`](https://www.npmjs.com/browse/depended/resume-schema) that actually import it in (many don't) have a section that looks like:
```js
function render() {
    try {
        return theme.render(resume);
    } catch (e) {
        console.log(e.message);
        return "";
    }
}
```
where "`resume`" is actually the schema and the themes mutate it with the data in the passed-in json resume object. See [`jsonresume-theme-elegant`](https://github.com/mudassir0909/jsonresume-theme-elegant/blob/master/render.js#L20-L27), [`jsonresume-theme-printclassy`](https://github.com/reubenkcoutinho/jsonresume-theme-printclassy/blob/master/server.js#L17-L25), [`jsonresume-theme-cora`](https://github.com/lechuckcaptain/jsonresume-theme-cora/blob/master/server.js#L17-L25) After checking about 50 themes, these three are the only ones I could find that depended on `resume-schema`. They don't use the validate export at all; I didn't look at all 300 or so themes, but the rest are many years old and target very old versions of `resume-schema` that won't be affected by anything going into `v1.0.0`.


> Submit changes to the validator under the resume-validator repo.

Are your most recent change in a yet-to-be pushed branch? What's on master errors like this:

```sh
> node index.js 
internal/modules/cjs/loader.js:670
    throw err;
    ^

Error: Cannot find module './schema'
```

**Comment by evanplaice** [comment]
Of course, Node assigns a specific meaning to `null` other than `null`. No wonder most JS devs dislike the existence of `null` in the language.

My bad. If z-shema makes the change to respect the 'null as first param rule' those changes will have to be added here as well. `undefined` or any value that indicates an invalid value will work.

---

I was thinking of going sync but we can keep the callback pattern. With that said, two options need to be added to the `args`.

**[ingnoreUnresolvableReferences](https://github.com/zaggino/z-schema#ignoreunresolvablereferences)**

```javascript
var validator = new ZSchema({
    ignoreUnresolvableReferences: true
});
```

There is an issue where a user reported that validation breaks without an internet connection. It does so b/c it's trying to validate this schema against the base JSONSchema definition but the link is broken w/ no internet.

**[breakOnFirstError](https://github.com/zaggino/z-schema#breakonfirsterror)**

```javascript
var validator = new ZSchema({
    breakOnFirstError: false
});
```

I can't think of any good reason why the validator should exit on the first error. The resume CLI should contain a list of *all* the errors if their resume has issues.

---

As for the validator. Props for doing the research. If the impact to dependents is small, we could just ditch the extra repo and migrate validation over to the CLI. The 1.0 goal is for `pkg.main` in this repo to point to the schema and all additional exports/scripts removed.

As for the `resume-validator` repo. It should contain the schema as a dependency and import from that. I've been intentionally neglecting that repo until I get a better idea of things. My preference is to delete it altogether.

**Comment by antialias** [comment]
> we could just ditch the extra repo and migrate validation over to the CLI

I put that one up yesterday: https://github.com/jsonresume/resume-cli/pull/354

**Comment by antialias** [comment]
> I can't think of any good reason why the validator should exit on the first error.

I think the reasoning here is that often the first error is the only sensical one; as an analogy consider language compilers, an unclosed brace can confuse the parser to the point where error messages don't make any sense.

**Comment by antialias** [comment]
> The resume CLI should contain a list of all the errors if their resume has issues.

Maybe we can avoid determining the best set of config options and instead read options set in `${process.cwd()}/jsonresume.config.js` that best suits whatever project the user is targeting.

**Comment by evanplaice** [comment]
FYI, as soon as the cleanup is done here I'll move on the the CLI (ie 1.0 schema can wait). You're a lot more familiar with the issues and PRs there so I'll be leaning on you for guidance.

The idea is to provide 'sane defaults'. Both options can be overridden with user-defined settings.

**Comment by antialias** [comment]
Should be fine as long as we have a good system for overriding the config values (e.g. `./jsonresume.config.js`.)

It seems both of these config options were added to `z-schema` to a few years ago in a way that they were opt-in, but since they are sensible options to to enable, I pitched them a couple pull requests to make it so:
https://github.com/zaggino/z-schema/pulls/antialias

My preference would be to avoid adding a layer of defaults to z-schema because _we_ think they are sane. It's more straightforward to defer our documentation to z-schema's defaults.

On a similar note, for where we do want to validate, we might consider switching to [`ajv`](https://github.com/epoberezkin/ajv). `ajv` is apparently [183 times faster than z-schema](https://rawgit.com/zaggino/z-schema/master/benchmark/results.html) by their own benchmark and also appears to be more actively maintained.

**Comment by antialias** [comment]
> My bad. If z-shema makes the change to respect the 'null as first param rule' those changes will have to be added here as well. undefined or any value that indicates an invalid value will work.

Unsure what your position is now. Shall I delete this branch or did you want to re-open and merge?

**Comment by evanplaice** [comment]
The 'unresolved deps' addresses this issue https://github.com/jsonresume/resume-cli/issues/268

Disabling break on first error removes the need for a -F flag

---

My position...

Move validation to resume-cli, which is already covered here https://github.com/jsonresume/resume-cli/pull/354

For cleanup. The validation API documentation needs to be removed from `README.md`. Removal of validation.js and the export will be covered under 1.0 planning.

**Comment by antialias** [comment]
>> `null` is used intentionally here to indicate that there is a value, it's just not a correctly typed valid value. `undefined` indicates the absence of a value. The tests can stay as-is.
> 
> BTW, the offending `undefined`s come from `z-schema` so I put up a PR that changes `undefined` to `null` in their codebase where it matters, [zaggino/z-schema#254](https://github.com/zaggino/z-schema/pull/254)
> 

Just FYI: https://github.com/zaggino/z-schema/pull/254 has been merged and published as 4.2.3.

**Comment by antialias** [comment]
(the initial concerns over `null` have been addressed completely with the z-schema update but keeping this closed to preserve @evanplaice's decision.)

**Comment by antialias** [comment]
> **[breakOnFirstError](https://github.com/zaggino/z-schema#breakonfirsterror)**
> 
> ```js
> var validator = new ZSchema({
>     breakOnFirstError: false
> });
> ```
> 
> I can't think of any good reason why the validator should exit on the first error. The resume CLI should contain a list of _all_ the errors if their resume has issues.

The `z-schema` maintainer appears to be in agreement with this as well, but they seem to think the `breakOnFirstError` is already the default even though it isn’t:
https://github.com/zaggino/z-schema/pull/255

**Comment by antialias** [comment]
> > **[breakOnFirstError](https://github.com/zaggino/z-schema#breakonfirsterror)**
> > ```js
> > var validator = new ZSchema({
> >     breakOnFirstError: false
> > });
> > ```
> > 
> > 
> > I can't think of any good reason why the validator should exit on the first error. The resume CLI should contain a list of _all_ the errors if their resume has issues.
> 
> The `z-schema` maintainer appears to be in agreement with this as well, but they seem to think that `breakOnFirstError:false` is already the default even though it isn’t:
> [zaggino/z-schema#255](https://github.com/zaggino/z-schema/pull/255)

[zaggino/z-schema#255](https://github.com/zaggino/z-schema/pull/255) merged and published as z-schema@5.0.0
